examples/shell/shell_common/cmd_protocol_server.cpp

```diff
@@ -0,0 +1,220 @@
+            mEchoServer.SetEchoRequestReceived(EchoServerCmd::HandleEchoRequestReceived);
+        }
+
+        ReturnErrorOnFailure(mSessionManager.NewPairing(peer, chip::kTestControllerNodeId, &mTestPairing,
+                                                        chip::SecureSessionMgr::PairingDirection::kResponder, kAdminId));
+        mConnected = true;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Stop()
+    {
+        if (!mConnected)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+        if (mTcp)
+        {
+            mTCPManager.Close();
+        }
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+        mUDPManager.Close();
+        mEchoServer.Shutdown();
+        mExchangeManager.Shutdown();
+        mSessionManager.Shutdown();
+        mAdmins.ReleaseAdminId(kAdminId);
+        mAdmins.Reset();
+        mConnected = false;
+
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    static void HandleEchoRequestReceived(chip::Messaging::ExchangeContext * ec, chip::System::PacketBufferHandle payload)
+    {
+        streamer_printf(streamer_get(), "Echo Request, len=%u ... sending response.\n", payload->DataLength());
+    }
+
+    EchoServerCmd(const EchoServerCmd &) = delete;
+    EchoServerCmd & operator=(const EchoServerCmd &) = delete;
+
+    bool mTcp       = false;
+    bool mConnected = false;
+
+    chip::Protocols::Echo::EchoServer mEchoServer;
+#if INET_CONFIG_ENABLE_IPV4
+    chip::TransportMgr<chip::Transport::UDP, chip::Transport::UDP> mUDPManager;
+#else
+    chip::TransportMgr<chip::Transport::UDP> mUDPManager;
+#endif // INET_CONFIG_ENABLE_IPV4
+
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    static constexpr size_t kMaxTcpActiveConnectionCount = 4;
+    static constexpr size_t kMaxTcpPendingPackets        = 4;
+#if INET_CONFIG_ENABLE_IPV4
+    chip::TransportMgr<chip::Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>,
+                       chip::Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>>
+        mTCPManager;
+#else  // INET_CONFIG_ENABLE_IPV4
+    chip::TransportMgr<chip::Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>> mTCPManager;
+#endif // INET_CONFIG_ENABLE_IPV4
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+    chip::SecureSessionMgr mSessionManager;
+    chip::SecurePairingUsingTestSecret mTestPairing;
+
+    chip::Messaging::ExchangeManager mExchangeManager;
+
+    chip::Transport::AdminPairingTable mAdmins;
+    chip::Transport::AdminPairingInfo * mAdminInfo     = nullptr;
+    static constexpr chip::Transport::AdminId kAdminId = 0;
+};
+
+EchoServerCmd gEchoServerCmd;
+
+} // namespace
+
+int cmd_protocol_server(int argc, char ** argv)
+{
+    bool tcp         = false;
+    bool stop        = false;
+    bool echoServer  = false;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (argc >= 1 && strcmp(argv[0], "stop") == 0)
```

Comment:
Maybe in the future consider getopt or argp.  This seems tedious for both authors and reviewers.